engine/execution/provider/engine.go

```diff
@@ -24,150 +28,222 @@ import (
+						Hex("chunk_id", logging.ID(chdpReq.ChunkID)).
+						Hex("requester_id", logging.ID(message.OriginID)).
+						Msg("chunk data pack request received")
+				}
+				return ok
+			},
+			// Map is called on messages that are Match(ed) successfully, i.e.,
+			// ChunkDataRequests.
+			// It replaces the payload of message with requested chunk id.
+			Map: func(message *engine.Message) (*engine.Message, bool) {
+				chdpReq := message.Payload.(*messages.ChunkDataRequest)
+				return &engine.Message{
+					OriginID: message.OriginID,
+					Payload:  chdpReq.ChunkID,
+				}, true
+			},
+			Store: chunkDataPackRequestQueue,
+		})
+
+	engine := Engine{
 		log:                    log,
 		tracer:                 tracer,
 		state:                  state,
-		me:                     me,
 		execState:              execState,
 		metrics:                metrics,
 		checkAuthorizedAtBlock: checkAuthorizedAtBlock,
-		chdpQueryTimeout:       time.Duration(chdpQueryTimeout) * time.Second,
-		chdpDeliveryTimeout:    time.Duration(chdpDeliveryTimeout) * time.Second,
+		chdpRequestHandler:     handler,
+		chdpRequestQueue:       chunkDataPackRequestQueue,
+		chdpRequestChannel:     make(chan *mempool.ChunkDataPackRequest, chdpRequestWorkers),
 	}
 
 	var err error
 
-	eng.receiptCon, err = net.Register(channels.PushReceipts, &eng)
+	engine.receiptCon, err = net.Register(channels.PushReceipts, &engine)
 	if err != nil {
 		return nil, fmt.Errorf("could not register receipt provider engine: %w", err)
 	}
 
-	chunksConduit, err := net.Register(channels.ProvideChunks, &eng)
+	chunksConduit, err := net.Register(channels.ProvideChunks, &engine)
 	if err != nil {
 		return nil, fmt.Errorf("could not register chunk data pack provider engine: %w", err)
 	}
-	eng.chunksConduit = chunksConduit
+	engine.chunksConduit = chunksConduit
+
+	cm := component.NewComponentManagerBuilder()
+	cm.AddWorker(engine.processQueuedChunkDataPackRequestsShovelerWorker)
+	for i := uint(0); i < chdpRequestWorkers; i++ {
+		cm.AddWorker(engine.processChunkDataPackRequestWorker)
+	}
 
-	return &eng, nil
+	engine.cm = cm.Build()
+	engine.Component = engine.cm
+
+	return &engine, nil
 }
 
-func (e *Engine) SubmitLocal(event interface{}) {
-	e.unit.Launch(func() {
-		err := e.ProcessLocal(event)
-		if err != nil {
-			engine.LogError(e.log, err)
+// processQueuedChunkDataPackRequestsShovelerWorker is constantly listening on the MessageHandler for ChunkDataRequests,
+// and pushes new ChunkDataRequests into the request channel to be picked by workers.
+func (e *Engine) processQueuedChunkDataPackRequestsShovelerWorker(ctx irrecoverable.SignalerContext, ready component.ReadyFunc) {
+	ready()
+
+	e.log.Debug().Msg("process chunk data pack request shovller worker started")
+
+	for {
+		select {
+		case <-e.chdpRequestHandler.GetNotifier():
+			// there is at list a single chunk data pack request queued up.
+			e.shovelChunkDataPackRequests()
+		case <-ctx.Done():
+			close(e.chdpRequestChannel)
```

Comment:
This means, if there are 500 request in the queue, and we shut down the engine, it won't stop until `shovelChunkDataPackRequests` finish processing all of the 500 requests in the queue. 

Why not dropping all the requests in the queue?

Reply:
cc @jordanschalm

Reply:
> if there are 500 request in the queue, and we shut down the engine, it won't stop until shovelChunkDataPackRequests finish processing all of the 500 requests in the queue.

Let's differentiate between the "message queue" and the "internal channel":
* the message queue is a HeroCache instance populated by incoming messages from the networking layer, it usually has a large size limit
* the internal channel (`chdpRequestChannel`) is a Go channel populated by the `processQueuedChunkDataPackRequestsShovelerWorker` goroutine. It can be much smaller (eg. currently it's size limit is [equal to the number of worker goroutines](https://github.com/onflow/flow-go/pull/2951/files#diff-99567a576e6710d390c77cd207573be724e407c3fab2d99f3cbd4bc6334e7572R116))

With the current implementation, we will wait for the smaller internal channel to drain, not the larger message queue. 

For context, [here](https://github.com/onflow/flow-go/pull/2951#discussion_r943982008) is the original comment suggesting that we drain the internal channel before stopping the engine. The reasoning for this was to make testing more responsive. Previously we needed to introduce a 1-second sleep into the test cases, because the test did not have a way to determine when a message was processed.

I agree, it is not desirable for an engine to continue processing message after it is stopped. Though personally I think the relatively small added time to stop the engine is worth the testing simplification.